src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/PoolingDataSource.cs

```diff
@@ -67,25 +336,111 @@ internal OpenInternalConnectionState(SqlConnectionX owningConnection, TimeSpan t
         }
 
         /// <inheritdoc/>
-        internal override ValueTask<SqlConnector> OpenNewInternalConnection(SqlConnectionX owningConnection, TimeSpan timeout, bool async, CancellationToken cancellationToken)
+        internal override ValueTask<SqlConnector?> OpenNewInternalConnection(SqlConnectionX owningConnection, TimeSpan timeout, bool async, CancellationToken cancellationToken)
         {
-            return _connectionRateLimiter.Execute<OpenInternalConnectionState, SqlConnector>(
+            return _connectionRateLimiter.Execute(
                 RateLimitedOpen,
                 new OpenInternalConnectionState(owningConnection, timeout),
                 async,
                 cancellationToken
             );
 
-            ValueTask<SqlConnector> RateLimitedOpen(OpenInternalConnectionState state, bool async, CancellationToken cancellationToken)
+            ValueTask<SqlConnector?> RateLimitedOpen(OpenInternalConnectionState state, bool async, CancellationToken cancellationToken)
             {
-                throw new NotImplementedException();
+                // As long as we're under max capacity, attempt to increase the connector count and open a new connection.
+                for (var numConnectors = _numConnectors; numConnectors < MaxPoolSize; numConnectors = _numConnectors)
+                {
+                    // Note that we purposefully don't use SpinWait for this: https://github.com/dotnet/coreclr/pull/21437
+                    if (Interlocked.CompareExchange(ref _numConnectors, numConnectors + 1, numConnectors) != numConnectors)
+                    {
+                        continue;
+                    }
+
+                    try
+                    {
+                        // We've managed to increase the open counter, open a physical connections.
+#if NET7_0_OR_GREATER
+                        var startTime = Stopwatch.GetTimestamp();
+#endif
+                        SqlConnector? connector = new SqlConnector(state._owningConnection, this);
+                        //TODO: set clear counter on connector
+
+                        //TODO: actually open the connector
+                        //await connector.Open(timeout, async, cancellationToken).ConfigureAwait(false);
```

Comment:
Should we not uncomment this, and add a `Task.Delay` in the connector.Open to simulate a delay ? I see your tests could benefit from this, instead of adding a sleep in your tests.

Reply:
I pushed a change to try this out, let's see how it goes in the pipeline 🤞

Reply:
This worked